internal/manifests/persistentvolume/persistentvolumeclaim.go

```diff
@@ -0,0 +1,103 @@
+package persistentvolume
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/ViaQ/logerr/kverrors"
+	"github.com/ViaQ/logerr/log"
+	"github.com/openshift/elasticsearch-operator/internal/manifests/status"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ComparePVCFunc is the type for functions that compare two persistentvolumeclaims.
+// Return true if two persistentvolumeclaim are equal.
+type ComparePVCFunc func(current, desired *corev1.PersistentVolumeClaim) bool
+
+// MutatePVCFunc is the type for functions that mutate the current persistentvolumeclaim
+// by applying the values from the desired persistentvolumeclaim.
+type MutatePVCFunc func(current, desired *corev1.PersistentVolumeClaim)
```

Comment:
can you explain the reasoning for introducing this instead of just comparing the differences of the objects (for example, we're provided the object we want to reflect the latest state) and just ensure that happens? (same ask for all the other k8s objects)

Reply:
The reasoning for separating compare and mutation is that we use a different set of comparison and mutation functions in different caller sites of `CreateOrUpdate`, e.g. deployments use a comparison function for ReplicaCount and a mutation function for setting the replica count while on a different place they compare the whole pod spec and pull in only dedicated chunks from desired to current spec.

In general the gist here is to come as close as possible to `controllerutil.CreateOrUpdate`. However as explained in the PR description the `controllerutil.CreateOrUpdate` assumes that resources provide the capability to be compared with `equality.Semantic.DeepEqual` and that is not applying to all `CreateOrUpdate` use cases in EO. Thus, I decided to keep a separate `compareFunc` for that reason.